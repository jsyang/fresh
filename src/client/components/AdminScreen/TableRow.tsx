import {Component, h} from 'preact';
import debounce from 'lodash.debounce';

import style from '../style';
import {Color} from '../colors';
import {getFormattedTaskValue} from './RoomSelect';
import {getRelevantTaskFields} from './SpreadSheet';
import {destroyTaskAdmin, getAllTasks, updateTaskAdmin} from '../../network/db';

const TableCell: any = style('td')({
    border:  `1px solid ${Color.Gray1}`,
    padding: '0.25em'
});

const getRelevantTaskFieldValues = task => Object.keys(task).filter(getRelevantTaskFields).map(key => task[key]);

export default class TableRow extends Component<{ row: any, headerKeys: string[] }> {
    onClickDelete = () => destroyTaskAdmin(this.props.row)
        .then(getAllTasks);

    onChange = () => {
        const {row} = this.props;

        const newRow: Partial<ITask> = {_id: row._id};

        Array.from(document.getElementById(row._id)!.querySelectorAll('[data-key]'))
            .forEach(fieldEl => {
                const key = fieldEl.getAttribute('data-key')!;

                let newValue;
                const textContent = fieldEl.firstChild!.textContent!.trim();

                switch (key) {
                    case'room':
                        newValue = (fieldEl.firstChild as any).value;
                        break;
                    case'isDeep':
                        newValue = textContent.toUpperCase() === 'Y';
                        break;
                    default:
                        newValue = textContent;
                        break;
                }

                newRow[key] = newValue;
            });

        updateTaskAdmin(newRow)
            .then(() => console.log('saved!'));
    };

    onChangeDebounced = debounce(this.onChange, 750);

    render() {
        const {row, headerKeys} = this.props;

        return <tr id={row._id}>
            {getRelevantTaskFieldValues(row).map((value, i) =>
                <TableCell
                    contenteditable={headerKeys[i] !== 'room'}
                    onBlur={this.onChangeDebounced}
                    onKeyUp={this.onChangeDebounced}
                    data-key={headerKeys[i]}
                >
                    {getFormattedTaskValue(value, headerKeys[i], this.onChangeDebounced)}
                </TableCell>
            )}
            <TableCell onClick={this.onClickDelete} style="cursor:pointer; text-align:center">‚ùå</TableCell>
        </tr>;
    }
}