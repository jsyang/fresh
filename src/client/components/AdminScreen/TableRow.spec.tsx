import {h} from 'preact';
import {render} from 'preact-render-to-string';

import TableRow from './TableRow';

describe('components/TableRow', () => {
    it('should match snapshot', () => {
        const renderable = (
            <TableRow
                row={{
                    _id:         'abcdef',
                    room:        'bed',
                    name:        'clean the floor',
                    description: 'floor gets really dirty after a party',
                    isDeep:      true
                }}
                headerKeys={'name,isDeep,description'.split(',')}
            />
        );

        expect(render(renderable)).toMatchSnapshot();
        expect(getCSSRules()).toMatchSnapshot('css');
    });
});
