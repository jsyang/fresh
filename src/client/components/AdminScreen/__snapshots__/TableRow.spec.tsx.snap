// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/TableRow should match snapshot 1`] = `

<tr id="abcdef">
  <td contenteditable
      data-key="name"
      class="p0"
  >
    bed
  </td>
  <td contenteditable
      data-key="isDeep"
      class="p0"
  >
    clean the floor
  </td>
  <td contenteditable
      data-key="description"
      class="p0"
  >
    Y
  </td>
  <td contenteditable
      class="p0"
  >
    floor gets really dirty after a party
  </td>
  <td style="cursor:pointer; text-align:center"
      class="p0"
  >
    ‚ùå
  </td>
</tr>

`;

exports[`components/TableRow should match snapshot: css 1`] = `
.p0 {
  border: 1px solid #aaaaaa;
  padding: 0.25em;
}
`;
